scylla-cdc-driver3/src/main/java/com/scylladb/cdc/cql/driver3/Driver3SchemaBuilder.java

```diff
@@ -12,125 +13,477 @@
+         * as of the time of CDC schema.
+         * <p>
+         * Proof of correctness under potential base table schema
+         * inconsistencies:
+         * <table border="1">
+         *     <caption>Proof of correctness</caption>
+         *     <tr>
+         *         <th>Inconsistency type</th>
+         *         <th>Schema change type</th>
+         *         <th>Commentary</th>
+         *     </tr>
+         *     <tr>
+         *         <td>Base schema newer than CDC log schema</td>
+         *         <td>Column added</td>
+         *         <td>This method will never be called for that added column,
+         *             as the column name was not present in the CDC log schema (caller).</td>
+         *     </tr>
+         *     <tr>
+         *         <td>Base schema newer than CDC log schema</td>
+         *         <td>Column removed</td>
+         *         <td>You cannot remove columns from the primary key. Therefore,
+         *             the removed column must have been a regular column. This
+         *             method will correctly return REGULAR for that column
+         *             (the type of that column as of the CDC schema time),
+         *             as it won't find this column in a primary key of the
+         *             base table.</td>
+         *     </tr>
+         *     <tr>
+         *         <td>Base schema newer than CDC log schema</td>
+         *         <td>Column renamed</td>
+         *         <td>Only clustering key columns can be renamed. We detect
+         *             if all clustering key columns in the base table
+         *             schema are present in the CDC log schema
+         *             (base schema: pk1, ck_new, col1;
+         *             CDC log schema: pk1, ck_old, col1). If there is
+         *             an inconsistency, that means that a clustering column
+         *             was renamed. That means that all other columns are not
+         *             changed. If the queried CDC column is a part of partition key
+         *             of base schema, that means that it was a partition
+         *             key (as of CDC schema time) - partition columns
+         *             were not modified (clustering column was modified).
+         *             Similarly for regular columns. But, if the column
+         *             is missing from the base table schema, that means
+         *             that we are seeing a old clustering column.
+         *             </td>
+         *     </tr>
+         *     <tr>
+         *         <td>Base schema newer than CDC log schema</td>
+         *         <td>Column data type changed</td>
+         *         <td>The column data type change does not change
+         *             the column set (column names, whether the columns
+         *             are in the primary key). Therefore, the schemas
+         *             are consistent with regard to column type.</td>
```

Comment:
I don't understand this one.
What does it mean that schemas are "consistent" when the column data type has changed?
See my previous comment - it's not clear what is actually being proven here (what the hypothesis is).

Reply:
Ahh, I think I understand. The definition of cosistency that you're using is as follows:
Let `s` be CDC log schema built by this instance of `Driver3SchemaBuilder`. Let `b(s)` be the corresponding base table schema. (note that we may not have access to `b(s)` - we may have a different version of the base table schema, older or newer). Then the column kind returned by this function is _consistent_ if it is the same as the column kind in `b(s)`.
I think this definition (or a variation of it - the wording may be different) should be precisely stated before you start proving anything.

Still, I don't know if it makes sense to talk about "consistency of schema". It could, if we extend the above definition from column kind to schemas, e.g.: "let `b*(s)` be the schema described by this instance of `BaseTableSchema`. We say that `b*(s)` is consistent if it is equal to `b(s)`". Where "equal" means that getting the column kinds, column types etc. through the accessor functions returns the same values as it would if we were dealing directly with `b(s)`. (observational equality)

What is **really** confusing in this entire proof and basically the entire file is the overloaded "column type". Just read this: "Column data type changed. ... Therefore the schemas are consistent with regard to column type". What the heck. But the truth is it should say: "Column data type changed. ... Therefore the schemas are consistent with regard to column **kind**" - and now it makes sense.

Reply:
Added a definition of inconsistency in the amended commit. By inconsistency, I mostly mean a situation where `ALTER TABLE` happened and one schema (base table or CDC log schema) is old (before `ALTER TABLE`) and the other one is new (after `ALTER TABLE`).